docs/test-suite/writing-portable-tests.md

```diff
@@ -178,7 +178,397 @@ This test _works_, but is _not_ very portable. If we move to a system with 192 c
+
+#### Inheriting from EESSI_Mixin
+
+The first step is to actually inherit from the `EESSI_Mixin` class:
+
+```
+from eessi.testsuite.eessi_mixin import EESSI_Mixin
+...
+@rfm.simple_test
+class EESSI_MPI4PY(rfm.RunOnlyRegressionTest, EESSI_Mixin):
+```
+
+#### Removing hard-coded test scales
+
+First, we remove 
+
+```
+    # ReFrame will generate a test for each scale
+    scale = parameter([2, 128, 256])
+```
+from the test. The `EESSI_Mixin` class will define the default set of scales on which this test will be run as
+```
+from eessi.testsuite.constants import SCALES
+...
+    scale = parameter(SCALES.keys())
+```
+
+This will ensure the test will run at all of the default scales, as defined by the `SCALES` [constant](https://github.com/EESSI/test-suite/blob/main/eessi/testsuite/constants.py).
+
+If, and only if, your test can not run on all of those scales should you overwrite this parameter in your child class. For example, if you have a test that does not support running on multiple nodes, you could define a filtering function outside of the class
+```
+def filter_scales():
+    return [
+        k for (k,v) in SCALES.items()
+        if v['num_nodes'] == 1
+    ]
+```
+and then in the class body overwrite the scales with a subset of items from the `SCALES` constant:
+```
+    scale = parameter(filter_scales())
+```
+
+Next, we also remove
+
+```
+   @run_after('init')
+    def define_task_count(self):
+        self.num_tasks = self.scale
+        self.num_tasks_per_node = min(self.num_tasks, 128)
+```
+
+as `num_tasks` and and `num_tasks_per_node` will be set by the `assign_tasks_per_compute_unit` [hook](https://github.com/EESSI/test-suite/blob/main/eessi/testsuite/hooks.py), which is invoked by the `EESSI_Mixin` class.
+
+Instead, we only set the `compute_unit`. A task will be launched for each compute unit. E.g.
+```
+    compute_unit = COMPUTE_UNIT[CPU]
+```
+will launch one task per (physical) CPU core. Other options are `COMPUTE_UNIT[HWTHREAD]` (one task per hardware thread), `COMPUTE_UNIT[NUMA_NODE]` (one task per numa node), `COMPUTE_UNIT[CPU_SOCKET]` (one task per CPU socket), `COMPUTE_UNIT[GPU]` (one task per GPU) and `COMPUTE_UNIT[NODE]` (one task per node). Check the `COMPUTE_UNIT` [constant](https://github.com/EESSI/test-suite/blob/main/eessi/testsuite/constants.py) for the full list of valid compute units. The number of cores per task will automatically be set based on this as the ratio of the number of cores in a node to the number of tasks per node (rounded down). The `EESSI_Mixin` class will also set the `OMP_NUM_THREADS` environment variable equal to the number of cores per task.
+
+!!! note
+    `compute_unit` needs to be set in ReFrame's `setup` phase (or earlier). For the different phases of the pipeline, please see the [documentation on how ReFrame executes tests](https://reframe-hpc.readthedocs.io/en/stable/pipeline.html).
```

Comment:
Not necessarily: one can still overwrite ReFrame's `setup` phase with one's own. So you could do something along the lines of:
```
def setup(self):
    compute_unit = 'something'
    super(self).setup()
```
and though I haven't tested this, I'm pretty sure it works. But sure: using pipeline hooks is the advocated way of doing things nowadays, so let's ignore this possibility for clarity.

Reply:
Oh, also note that I used this phrasing since it's the same as we use in our `validate_init` check: https://github.com/EESSI/test-suite/blob/caf5391449517cd49ba614e8be9e5fa967ccb700/eessi/testsuite/eessi_mixin.py#L87

Actually, I'll not commit this change for now, let's first make sure we're on the same page on what phrasing we want to keep. I do think the current phrasing is more _exact_. I.e. _if_ one has the rare case where they overwrite the init of setup phase, they know they are still allowed to set it there. And who knows, there might be very particular cases where you need this. For example, if you _need_ the setup phase to have run in order to have the right information, one could still do
```
def setup(self):
    super(self).setup()
    compute_unit = <something_depending_on_current_state_of_self>
```
That would still set the `compute_unit` _before_ the `run_after('setup')` hook is called by the `EESSI_Mixin` class. Note that this is one way in which we could resolve the 'problem' that `EESSI_Mixin` hooks always run before the child class' hooks.

Reply:
ok, fair point.
the recommended way is indeed to use the pipeline hooks, so how about rephrasing it so the `before` comes first?
```
`compute_unit` needs to be set _before_ (or in) ReFrame's `setup` phase.
```

Reply:
Perfectly fine by me. I'll rephrase here, and try to also make a PR to the test suite to swap the order there as well.